server/aws-lsp-codewhisperer/src/language-server/chat/contexts/documentFqnExtractor.ts

```diff
@@ -0,0 +1,161 @@
+import { DocumentSymbol, SymbolType } from '@amzn/codewhisperer-streaming'
+import { FqnExtractorOutput, FullyQualifiedName, fqnExtractor } from '@aws/lsp-fqn'
+import { Range, TextDocument } from 'vscode-languageserver-textdocument'
+import { Features, Result } from '../../types'
+
+export interface DocumentFqnExtractorConfig {
+    nameMinLength?: number
+    nameMaxLength?: number
+    maxSymbols?: number
+    timeout?: number
+    logger?: Features['logging']
+}
+
+export type FqnSupportedLanagues =
+    | 'python'
+    | 'java'
+    | 'javascript'
+    | 'javascriptreact'
+    | 'typescript'
+    | 'typescriptreact'
+
+export class DocumentFqnExtractor {
+    public static readonly DEFAULT_CONFIG = {
+        nameMinLength: 1,
+        nameMaxLength: 256,
+        maxSymbols: 25,
+    }
+
+    public static readonly FQN_SUPPORTED_LANGUAGE_SET = new Set([
+        'python',
+        'java',
+        'javascript',
+        'typescript',
+        'javascriptreact',
+        'typescriptreact',
+    ])
+
+    #nameMinLength: number
+    #nameMaxLength: number
+    #maxSymbols: number
+    #timeout?: number
+
+    #logger?: Features['logging']
+    #disposeFunctionSet: Set<() => void>
+
+    constructor(config?: DocumentFqnExtractorConfig) {
+        const { nameMaxLength, nameMinLength, maxSymbols, timeout, logger } = config ?? {}
+        this.#nameMinLength = nameMinLength ?? DocumentFqnExtractor.DEFAULT_CONFIG.nameMinLength
+        this.#nameMaxLength = nameMaxLength ?? DocumentFqnExtractor.DEFAULT_CONFIG.nameMaxLength
+        this.#maxSymbols = maxSymbols ?? DocumentFqnExtractor.DEFAULT_CONFIG.maxSymbols
+        this.#timeout = timeout
+        this.#logger = logger
+        this.#disposeFunctionSet = new Set()
+    }
+
+    public dispose() {
+        this.#disposeFunctionSet.forEach(disposeFunction => disposeFunction())
```

Comment:
When will those `disposeFunction`'s be called? I see a long chain of objects creation and it's difficult to understand when stuff is going to be disposed. I just try to verify we do not create memory issues with this.

Reply:
In the way it is structured, there could be multiple threads spawned from this class at the same time. - I think I will change it so we instantiate a new instance every time so we have a more straightforward logic.

Reply:
Workerpool simplified this, so no longer relevant